File: haskell/2023/Day9.hs
module Day9 where

import Data.List.Split

asPairs :: [a] -> [(a, a)]
asPairs [l, r] = [(l, r)]
asPairs (l : r : xs) = (l, r) : asPairs (r : xs)

process1 :: [Int] -> Int
process1 = sum . go []
  where
    go acc xs = if all (== 0) xs' then acc' else go acc' xs'
      where
        acc' = acc <> [last xs]
        xs' = map (\(l, r) -> r - l) $ asPairs xs

process2 :: [Int] -> Int
process2 = foldr1 (-) . go []
  where
    go acc xs = if all (== 0) xs' then acc' else go acc' xs'
      where
        acc' = acc <> [head xs]
        xs' = map (\(l, r) -> r - l) $ asPairs xs

main' :: IO ()
main' = do
  contents <- map (map read . splitOn " ") . lines <$> readFile "../inputs/2023/Day9/input.txt" :: IO [[Int]]
  -- part 1
  print $ sum $ map process1 contents
  -- part 2
  print $ sum $ map process2 contents
  return ()

File: typescript/2023/day9.ts
import { aperture, head, last, sum } from 'ramda';

const content = (await Bun.file('../inputs/2023/Day9/input.txt').text())
  .trim()
  .split('\n')
  .map(row => row.split(' ').map(s => Number.parseInt(s, 10)));

export const process1 = (history: number[]) => {
  const acc: number[] = [];
  let arr = history;

  while (!arr.every(v => v === 0)) {
    acc.push(last(arr)!);
    arr = aperture(2, arr).map(([l, r]) => r - l);
  }

  return sum(acc);
};

const result1 = sum(content.map(process1));

console.log(result1);

export const process2 = (history: number[]) => {
  const acc: number[] = [];
  let arr = history;

  while (!arr.every(v => v === 0)) {
    acc.push(head(arr)!);
    arr = aperture(2, arr).map(([l, r]) => r - l);
  }

  return acc.reduceRight((l, r) => r - l);
};

const result2 = sum(content.map(process2));
console.log(result2);
